flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/CassandraSource.java

```diff
@@ -0,0 +1,162 @@
+    private static final long serialVersionUID = 7773196541275567433L;
+
+    private final ClusterBuilder clusterBuilder;
+    private final Class<OUT> pojoClass;
+    private final String query;
+    private final MapperOptions mapperOptions;
+
+    public CassandraSource(
+            ClusterBuilder clusterBuilder,
+            Class<OUT> pojoClass,
+            String query,
+            MapperOptions mapperOptions) {
+        checkNotNull(clusterBuilder, "ClusterBuilder required but not provided");
+        checkNotNull(pojoClass, "POJO class required but not provided");
+        checkQueryValidity(query);
+        this.clusterBuilder = clusterBuilder;
+        ClosureCleaner.clean(clusterBuilder, ExecutionConfig.ClosureCleanerLevel.RECURSIVE, true);
+        this.pojoClass = pojoClass;
+        this.query = query;
+        this.mapperOptions = mapperOptions;
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return Boundedness.BOUNDED;
+    }
+
+    @Internal
+    @Override
+    public SourceReader<OUT, CassandraSplit> createReader(SourceReaderContext readerContext) {
+        return new CassandraSourceReader<>(
+                readerContext, clusterBuilder, pojoClass, query, mapperOptions);
+    }
+
+    @Internal
+    @Override
+    public SplitEnumerator<CassandraSplit, CassandraEnumeratorState> createEnumerator(
+            SplitEnumeratorContext<CassandraSplit> enumContext) {
+        return new CassandraSplitEnumerator(enumContext, null, clusterBuilder);
+    }
+
+    @Internal
+    @Override
+    public SplitEnumerator<CassandraSplit, CassandraEnumeratorState> restoreEnumerator(
+            SplitEnumeratorContext<CassandraSplit> enumContext,
+            CassandraEnumeratorState enumCheckpoint) {
+        return new CassandraSplitEnumerator(enumContext, enumCheckpoint, clusterBuilder);
+    }
+
+    @Internal
+    @Override
+    public SimpleVersionedSerializer<CassandraSplit> getSplitSerializer() {
+        return CassandraSplitSerializer.INSTANCE;
+    }
+
+    @Internal
+    @Override
+    public SimpleVersionedSerializer<CassandraEnumeratorState> getEnumeratorCheckpointSerializer() {
+        return CassandraEnumeratorStateSerializer.INSTANCE;
+    }
+
+    @Override
+    public TypeInformation<OUT> getProducedType() {
+        return TypeInformation.of(pojoClass);
+    }
+
+    @VisibleForTesting
+    public static void checkQueryValidity(String query) {
+        checkNotNull(query, "query required but not provided");
+        checkState(
+                query.matches(CassandraSplitReader.SELECT_REGEXP),
+                "query must be of the form select ... from keyspace.table ...;");
+        checkState(
+                !query.matches(CQL_PROHIBITTED_CLAUSES_REGEXP),
+                "query must not contain aggregate or order clauses because they will be done per split. "
+                        + "So they will be incorrect after merging the splits");
```

Comment:
users wont understand why splits are being merged,

Reply:
I don't get what you mean. What I meant here is that any oder or grouping SQL operation will be done at the split level so they will be incorrect at the DataStream level. So I chose to disable them in the source and let the user use the regular methods in DataStream instead.

Reply:
This is purely about phrasing.
"Splits" is ultimately a source-internal concept that no one that hasn't written their own source is aware of.
I'd suggest something like "Aggregations/OrderBy are not supported because the query is executed on subsets/partitions of the input table."

Note that it _can_ still be useful to allow aggregation or order clauses. For example, pushing an aggregate into cassandra and applying a aggregate function in Flink _can_ be more efficient than fetching everything from Cassandra and doing all of the aggregation in Flink.

If we want to support Flink SQL we'll need to revisit this, because it would be nice to to support pushdown aggregates. 

On that note, how does groupBy interact with token ranges? I assume there's no guarantee that a particular subtask gets all values for a given key, because in practice the token ranges split the input into effectively random partitions?

Reply:
Ok I'll rephrase to your suggestion. 
Regarding pushdown I agree, it will be needed in the future but is out of the scope of the current PR. So I suggest we keep aggregates unsupported in this PR and open a ticket to keep track of this future feature.